src/core/packet.lua

```diff
@@ -20,10 +18,40 @@ local header_size = 8
 local max_payload = tonumber(C.PACKET_PAYLOAD_SIZE)
 
 -- Freelist containing empty packets ready for use.
+
+ffi.cdef[[
+struct freelist {
+    int nfree, max;
+    struct packet *list[?];
+};
+]]
```

Comment:
Thanks for doing the refactor, looks great :)  I have two more nits before landing:

1. Would you mind changing to unsigned integers please, unless you have a reason to use signed integers
2. I wonder about "struct freelist" -- I can see this name as easily colliding with a name from some other user's module, struct tags being global and all.  Unless you need this struct to have a tag, in this case "local freelist_t = ffi.typeof[[struct { ... }]] is more appropriate.  In that case I would change the ffi.new call below to be "local packets_fl = freelist_t(max_packets, 0, max_packets).

Thanks!